## Задание 1

### Считать данные с помощью pandas

In [2]:
import pandas as pd

In [5]:
df = pd.read_csv('kc_house_data.csv')

### Вывести на экран первые 5 строк

In [7]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7      1180.0              0   
1      7242     2.0           0     0  ...      7      2170.0            400   
2     10000     1.0           0     0  ...      6       770.0              0   
3      5000     1.0           0     0  ...      7      1050.0            910   
4      8080     1.0           0     0  ...      8      1680.0              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### Посмотреть на описание признаков и на их содержание

В базе данных есть возможность сортировки по id, дате, цене, количеству спален, санузлов, площади, этажности, виду, оценке, площади  подвала, году постройки и реновации, индексу, координатам.

## Задание 2

### Проведите первичный анализ данных. Изучите типы данных.

In [16]:
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above       float64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

По всей видимости мы загрузили базу недвижимости США. Площадь измеряется в квадратных футах (int64), санузлы описываются типом данных float64, их принято делить на четверти в зависимости от наполнения. Дата записана типом object, то есть это текст.

### Найдите количество пропущенных ячеек в данных

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21611 non-null  float64
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

База данных полная, пропущенных ячеек нет.

### Посчитайте основные статистики по всем признакам и поизучайте их. Пишите выводы.


In [20]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21611.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.396095     291.509045   1971.005136     84.402258   
std        1.175459    828.128162     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

Цены на жилье варьируются от 75000 до 7700000, средняя цена около 540000. Медианная цена 450000.

В домах от нуля до 33 спален. Медианное значение - 3.

В домах от нуля до 8 санузлов. Медианное значение - 2.25.

Жилая площадь от 290 до 13540 квадратных футов. Медианное значение 1910.

Площадь участка от 520 до более чем 1650000 квадратных футов. Медианное значение 7618, среднее 1510.

В домах от 1 до 3.5 этажей, медианное значение 1.5.

Есть возможность учитывать, стоит ли дом на берегу водоема (да или нет).

Колонка view по всей видимости означает ценность вида из окон по шкале от 0 до 4.

Состояние домов оценивается по пятибалльной шкале, средняя оценка 3.409, медианная 3.

Колонка grade вводит какую-то шкалу оценок от 1 до 13 с медианной оценкой 7 баллов.

Колонки sqft_above и sqft_basement позволяют оценить соотношение подвала к остальной олощади дома. Площадь подвалов от 0 (его нет) до 4820 кв.футов. Медианное значение площади подвала 0, то есть в большинстве домов его нет.

Колонки yr_built и yr_renovated хранят информацию о том, в каком году был дом построен и отреставрирован. Дома от 1900 до 2015 года постройки, медианное значение 1997, среднее 1971. В колонке  yr_renovated 0 означает отсутствие капитального ремонта, поэтому среднее значение не информативно. Медианное значение 0 говорит, что в большинстве домов капитального ремонта не было.

Колонка zipcode хранит индекс.

Остальные колонки пока не понятны.

## Задание 3

### 3.1 В каком диапазоне изменяются стоимости недвижимости?

In [21]:
df['price'].min()

75000.0

In [22]:
df['price'].max()

7700000.0

От 75000 до 7700000.

### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

*В базе нет разделения на жилую/не жилую площадь, скорее всего имеется в виду sqft_living и sqft_lot - жилой площади к площади участка.*

In [24]:
df['sqft_living'].mean()

2079.8997362698374

In [25]:
df['sqft_lot'].mean()

15106.967565816869

In [27]:
2079/15106*100

13.76274328081557

### 3.3 Как много домов с разными этажами в данных?

In [28]:
df['floors'].unique()

array([1. , 2. , 1.5, 3. , 2.5, 3.5])

Встречаются дома от 1 до 3.5 этажей с шагом в 0.5 этажа. Всего 6 типов.

### 3.4 Насколько хорошие состояния у домов в данных?

In [29]:
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

Большинство домов оценено 3 баллами.

### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [30]:
df['yr_built'].min()

1900

In [31]:
df['yr_built'].max()

2015

## Задание 4

### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [51]:
bedrooms2 = df[df['bedrooms'] == 2]

In [53]:
bedrooms2['price'].mean()

401372.681884058

### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [56]:
over600000 = df[df['price'] >= 600000]

In [57]:
over600000['sqft_lot'].mean()

20454.783537543197

### 4.3 Как много домов коснулся ремонт?

In [60]:
df[df['yr_renovated'] > 0]

id             date      price  bedrooms  bathrooms  \
1      6414100192  20141209T000000   538000.0         3       2.25   
35     9547205180  20140613T000000   696000.0         3       2.50   
95     1483300570  20140908T000000   905000.0         4       2.50   
103    2450000295  20141007T000000  1088000.0         3       2.50   
115    3626039325  20141121T000000   740500.0         3       3.50   
...           ...              ...        ...       ...        ...   
19622  7351200295  20150114T000000  1150000.0         3       1.75   
20057   126039256  20140904T000000   434900.0         3       2.00   
20444  4305600360  20150225T000000   500012.0         4       2.50   
20447  3319500628  20150212T000000   356999.0         3       1.50   
20962  1278000210  20150311T000000   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1             2570      7242     2.0           0     0  ...      7   
35            2300      3060     1.5           0     0  ...      8   
95            3300     10250     1.0           0     0  ...      7   
103           2920      8113     2.0           0     0  ...      8   
115           4380      6350     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
19622         1760      6788     2.0           1     4  ...      7   
20057         1520      5040     2.0           0     0  ...      7   
20444         2400      9612     1.0           0     0  ...      8   
20447         1010      1546     2.0           0     0  ...      8   
20962          828      4524     1.0           0     0  ...      6   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1          2170.0            400      1951          1991    98125  47.7210   
35         1510.0            790      1930          2002    98115  47.6827   
95         2390.0            910      1946          1991    98040  47.5873   
103        2920.0              0      1950          2010    98004  47.5814   
115        2780.0           1600      1900          1999    98117  47.6981   
...           ...            ...       ...           ...      ...      ...   
19622      1760.0              0      1940          1960    98125  47.7336   
20057      1520.0              0      1977          2006    98177  47.7770   
20444      1230.0           1170      1962          2009    98059  47.4799   
20447      1010.0              0      1971          2014    98144  47.5998   
20962       828.0              0      1968          2007    98001  47.2655   

          long  sqft_living15  sqft_lot15  
1     -122.319           1690        7639  
35    -122.310           1590        3264  
95    -122.249           1950        6045  
103   -122.196           2370        8113  
115   -122.368           1830        6350  
...        ...            ...         ...  
19622 -122.284           1630        7588  
20057 -122.362           1860        8710  
20444 -122.127           2430        5539  
20447 -122.311           1010        1517  
20962 -122.244            828        5402  

[914 rows x 21 columns]

### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [63]:
grade10 = df[df['grade'] > 10]

In [64]:
grade10['price'].mean()

1678635.1175298805

In [66]:
grade4 = df[df['grade'] < 4]

In [67]:
grade4['price'].mean()

189750.0

In [69]:
1678635.1175298805-189750.0

1488885.1175298805

## Задание 5

### 5.1 Выберите дом клиенту. Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?


In [71]:
df[(df['waterfront'] == 1) & (df['bedrooms'] >= 3) & (df['sqft_basement'] > 0)]

id             date      price  bedrooms  bathrooms  \
49      822039084  20150311T000000  1350000.0         3       2.50   
246    2025069065  20140929T000000  2400000.0         4       2.50   
300    3225069065  20140624T000000  3075000.0         4       5.00   
457    2122039094  20141126T000000   705000.0         3       3.00   
656    3760500116  20141120T000000  3070000.0         3       2.50   
...           ...              ...        ...       ...        ...   
18914  9253900408  20150408T000000  1400000.0         3       2.75   
20325   518500480  20140811T000000  3000000.0         3       3.50   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
49            2753     65005     1.0           1     2  ...      9   
246           3650      8354     1.0           1     4  ...      9   
300           4550     18641     1.0           1     4  ...     10   
457           1970     20978     2.0           1     3  ...      9   
656           3930     55867     1.0           1     4  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
18914         3130     19530     1.0           1     4  ...      8   
20325         4410     10756     2.0           1     4  ...     11   
20767         4360      8175     2.5           1     4  ...     10   
21201         3760      5634     2.0           1     4  ...     11   
21576         4850     10584     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
49         2165.0            588      1953             0    98070  47.4041   
246        1830.0           1820      2000             0    98074  47.6338   
300        2600.0           1950      2002             0    98074  47.6053   
457        1770.0            200      1980             0    98070  47.3844   
656        2330.0           1600      1957             0    98034  47.7022   
...           ...            ...       ...           ...      ...      ...   
18914      1690.0           1440      1947          1984    98008  47.5895   
20325      3430.0            980      2014             0    98056  47.5283   
20767      3940.0            420      2007             0    98008  47.5724   
21201      2830.0            930      2014             0    98056  47.5285   
21576      3540.0           1310      2007             0    98008  47.5943   

          long  sqft_living15  sqft_lot15  
49    -122.451           2680       72513  
246   -122.072           3120       18841  
300   -122.077           4550       19508  
457   -122.438           2280       75396  
656   -122.224           2730       26324  
...        ...            ...         ...  
18914 -122.111           2980       18782  
20325 -122.205           3550        5634  
20767 -122.104           2670        8525  
21201 -122.205           3560        5762  
21576 -122.110           3470       18270  

[84 rows x 21 columns]

### 5.2 Выберите дом клиенту. Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?


In [78]:
case2 = df[(df['view'] == 4) | (df['waterfront'] == 1) & (df['grade'] > 11) & (df['yr_built'] >= 1980)]

In [79]:
case2['price'].min()

252000.0

In [80]:
case2['price'].max()

6885000.0

### 5.3 Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [81]:
case3 = df[(df['sqft_basement'] == 0) & (df['floors'] == 2.0) & (df['price'] < 150000.0)]

In [83]:
case3['condition'].mean()

2.8333333333333335